### DISCUSSION TITLE
How to change time varying parameters on runtime

### DISCUSSION THREAD

**Discussion by gmsanchez** [do-mpc/do-mpc#240] [Q&A] [answered]
Hi!

I am following the [TVP FAQ](https://www.do-mpc.com/en/latest/FAQ.html#time-varying-parameters) and [Issue #92](https://github.com/do-mpc/do-mpc/issues/92) code regarding the use of TVP. However it is not clear to me how to change the parameters values during runtime. Let's suppose that I want to change the `temperature_set_point` arbitrarily without knowing the iteration step `k`. How can I implement that?

Thanks in advance

**Comment by 4flixt** [comment, marked as answer]
Hi @gmsachez,

it is really quite easy: The ``tvp_fun``which is set to the MPC controller with ``set_tvp_fun`` is called at each new time-instance. In this function, whatever values you write onto the ``tvp_template`` (the mandatory return item of the ``tvp_fun``) will be considered as the currently valid time-varying parameters when solving the optimal control problem.

Hope This answers your question.

Best,
4flixt

**Reply by HosameldinMohamed** [reply]
Hi @4flixt,

I see that the `tvp_fun` allows only `t_now` to be the input. And is intended to be defined in optimizer configuration.

So it's possible to set the time-varying parameters using a predefined sequence. But is it possible to set a parameter with a value that's known only at a particular time instance (inside the loop where we call `make_step()`)?

**Reply by 4flixt** [reply]
Hi @HosameldinMohamed,

it is in fact quite easy. I often do something like this:

```python
tvp_template = mpc.get_tvp_template()
# Create dummy function simply returning the tvp_template structure
tvp_fun = lambda t_now: tvp_template
mpc.set_tvp_fun(tvp_fun)

# In the mainloop:
tvp_template[ ... ] = ... # (set whatever value, e.g. taken from simulator)
u0 = mpc.make_step(x0)
```

Hope this helps!

Best,
4flixt

**Reply by HosameldinMohamed** [reply]
Many thanks for your fast reply @4flixt!
I just tried it and it works!
And thanks for the toolbox!!
